### PR TITLE
Feature/textfield custom validity

### PR DESCRIPTION
**Pull request by Teoxoy**
https://github.com/material-components/material-components-web/issues/1018 - Added new feature to MDCTextfield

#1053 - moved here

### CODE REVIEW

File: packages/mdc-textfield/README.md [hunk 1]
```diff
@@ -153,9 +153,9 @@ by HTML5's form validation API.
 </div>
 ```
 
-An input's validity is checked via `checkValidity()` on blur, and the styles are updated
-accordingly. When using the `required` attribute, an asterisk will be automatically appended to the
-label text, as per the spec.
+By default an input's validity is checked via `checkValidity()` on blur, and the styles are updated
+accordingly. The validity can also be set via the MDCTextfield.valid variable. When using the `required`
```
Comment (lynnjepsen) [inline packages/mdc-textfield/README.md]: Try this:

Set the MDCTextfield.valid variable to set the input's validity explicitly. MDC Textfield automatically appends an asterisk to the label text if the `required` attribute is set.

File: packages/mdc-textfield/foundation.js [hunk 2]
```diff
@@ -99,7 +100,7 @@ export default class MDCTextfieldFoundation extends MDCFoundation {
   }
 
   deactivateFocus_() {
-    const {FOCUSED, INVALID, LABEL_FLOAT_ABOVE} = MDCTextfieldFoundation.cssClasses;
+    const {FOCUSED, LABEL_FLOAT_ABOVE} = MDCTextfieldFoundation.cssClasses;
     const input = this.getNativeInput_();
     const isValid = input.checkValidity();
```
Comment (lynnjepsen) [inline packages/mdc-textfield/foundation.js]: can you change the implementation of the `checkValidity` method to check `useNativeValidityChecking_`?

It's odd that `deactivateFocus_` would both `checkValidity` and `changeValidity`

File: packages/mdc-textfield/foundation.js [hunk 3]
```diff
@@ -167,4 +175,9 @@ export default class MDCTextfieldFoundation extends MDCFoundation {
       badInput: false,
     };
   }
+
+  setValid(isValid) {
+    this.useNativeValidityChecking_ = false;
```
Comment (lynnjepsen) [inline packages/mdc-textfield/foundation.js]: I think there should also be a way to go back to using native validity checking

File: packages/mdc-textfield/foundation.js [hunk 4]
```diff
@@ -57,6 +57,7 @@ export default class MDCTextfieldFoundation extends MDCFoundation {
     this.inputBlurHandler_ = () => this.deactivateFocus_();
     this.inputInputHandler_ = () => this.autoCompleteFocus_();
     this.inputKeydownHandler_ = () => this.receivedUserInput_ = true;
+    this.useNativeValidityChecking_ = true;
```
Comment (lynnjepsen) [inline packages/mdc-textfield/foundation.js]: the word "native" here is a little misleading. Makes it sound like its happening in an Android/iOS app, not web.

I think what you mean is that MDC Textfield has some default validity checking, and this variable overrides that.

why not `useCustomValidityChecking_` and flip true to false?

Comment (kfranqueiro) [inline packages/mdc-textfield/foundation.js]: "Native" here refers to depending on the browser's native HTML and DOM features for form validation; I think that's reasonable, but flipping the variable name is also reasonable.

Comment (Teoxoy) [inline packages/mdc-textfield/foundation.js]: Yes, that's why I used "Native". Should I leave it as is or modify it to "Custom" ?

Comment (kfranqueiro) [inline packages/mdc-textfield/foundation.js]: IMO "Custom" and defaulting to `false` might be easier to understand anyway. i.e. the default behavior is no custom validity (use native HTML validity checking).  The non-default is adding something moreso than removing something.

File: packages/mdc-textfield/foundation.js (hunk 2 above)
Comment (Teoxoy) [inline packages/mdc-textfield/foundation.js]: `checkValidity` is a native HTML/JS function for validating the input.

Comment (Teoxoy) [inline packages/mdc-textfield/foundation.js]: I will move the input check in the if block just for performance reasons like so:
`if (!this.useCustomValidityChecking_) {
  this.changeValidity_(input.checkValidity());
}`

File: packages/mdc-textfield/foundation.js (hunk 3 above)
Comment (Teoxoy) [inline packages/mdc-textfield/foundation.js]: I'm not sure if someone would ever use that feature. Is it worth having another public variable to change that ?

File: packages/mdc-textfield/foundation.js (hunk 2 above)
Comment (kfranqueiro) [inline packages/mdc-textfield/foundation.js]: I may be missing something, but shouldn't we not even compute `isValid` if not using native validity checking? Shouldn't that be moved inside the `if` as well?

File: packages/mdc-textfield/foundation.js (hunk 3 above)
Comment (kfranqueiro) [inline packages/mdc-textfield/foundation.js]: I'm curious of a use-case for that as well. I would think it's not typical (and potentially confusing) to change validation logic of a particular component instance on-the-fly, so you're either going to use this method directly to control valid state yourself, or you're going to rely on native validity checking.

**Comment by Teoxoy** [comment]
I stated this on the issue #1018 : 

> The valid variable is just a setter. The only way to implement a getter in the foundation would require the adapter to have a function hasClass and check for this.adapter_.hasClass(INVALID); or an additional variable that keeps track of the validity.

Would it be worth to implement a getter ?
IMO if the user sets the variable, he will not need to check it.

File: test/unit/mdc-textfield/foundation.test.js [hunk 5]
```diff
@@ -246,19 +258,34 @@ test('on blur does not remove mdc-textfield__label--float-above if input has a v
   td.verify(mockAdapter.removeClassFromLabel(cssClasses.LABEL_FLOAT_ABOVE), {times: 0});
 });
 
-test('on blur removes mdc-textfield--invalid if input.checkValidity() returns true', () => {
+test('on blur removes mdc-textfield--invalid if setValid was not set and input.checkValidity() returns true', () => {
```
Comment (lynnjepsen) [inline test/unit/mdc-textfield/foundation.test.js]: if custom validity is false

File: test/unit/mdc-textfield/foundation.test.js [hunk 6]
```diff
@@ -246,19 +258,34 @@ test('on blur does not remove mdc-textfield__label--float-above if input has a v
...
   const {mockAdapter, blur} = setupBlurTest();
   blur();
   td.verify(mockAdapter.removeClass(cssClasses.INVALID));
 });
 
-test('on blur adds mdc-textfied--invalid if input.checkValidity() returns false', () => {
+test('on blur adds mdc-textfied--invalid if setValid was not set and input.checkValidity() returns false', () => {
```
Comment (lynnjepsen) [inline test/unit/mdc-textfield/foundation.test.js]: if custom validity is false

File: test/unit/mdc-textfield/foundation.test.js [hunk 7]
```diff
@@ -246,19 +258,34 @@ test('on blur does not remove mdc-textfield__label--float-above if input has a v
...
   const {mockAdapter, blur, nativeInput} = setupBlurTest();
   nativeInput.checkValidity = () => false;
   blur();
   td.verify(mockAdapter.addClass(cssClasses.INVALID));
 });
 
+test('on blur does not remove mdc-textfield--invalid if setValid was set and input.checkValidity() returns true', () => {
```
Comment (lynnjepsen) [inline test/unit/mdc-textfield/foundation.test.js]: if custom validity is true

File: test/unit/mdc-textfield/foundation.test.js [hunk 8]
```diff
@@ -246,19 +258,34 @@ test('on blur does not remove mdc-textfield__label--float-above if input has a v
...
+  const {foundation, mockAdapter, blur} = setupBlurTest();
+  foundation.setValid(false);
+  blur();
+  td.verify(mockAdapter.removeClass(cssClasses.INVALID), {times: 0});
+});
+
+test('on blur does not add mdc-textfied--invalid if setValid was set and input.checkValidity() returns false', () => {
```
Comment (lynnjepsen) [inline test/unit/mdc-textfield/foundation.test.js]: if custom validity is true

**Comment by lynnjepsen** [comment]
> Would it be worth to implement a getter ? IMO if the user sets the variable, he will not need to check it.

I think you're right. No need to check the variable if the user sets it.

**Comment by lborgman** [comment]
I just wanted to read the value of ".valid". It took me some time to find out why it didn't work. 😐

I think it's quite confusing to have a setter and a non-functional getter. (I still have not read the spec, but I found this )

My usecase is just something similar to a form validation before submit.

**Pull request merged**